### PR TITLE
Refactors the list datatype to support composite lists. Adapts a lot of circuits to be able to properly use composite lists. Adds the dispenser shell

### PR DESCRIPTION
**Pull request by Watermelon914**
## About The Pull Request

The dispenser shell lets you dispense items, simple as that. It uses an entity list to let you vend whatever item you've put in. Can only accept items of size normal or lower and has a limit of 20 items.

Adds composite datatypes and with it comes composite lists. All lists have been converted to be composite lists and adds some extra list operation components to iterate through each list and also filter the list. A lot of refactoring needed to be done to add this composite datatype and assoc list had to also be moved to be a composite datatype.
NTNet components now send and receive lists of different types, with the receiver only accepting the value if the types match. This makes sending and receiving data with NTNet components slightly easier.

I started refactoring variables in this PR as I was preparing to also add basic list manipulation but decided the PR was getting too big. The variable refactors have been left in for when I get to doing basic list manipulation.

Two new components that were added:
- Filter list - Filters a list using specific conditions. During the testing of this component, I found out that if you chain multiple instant executing circuits, it would runtime and cause bugs so this was also fixed for the filter list component.
- For each component - Iterates through the list and outputs each entity

Fixes some general bugs with some of the list components for integrated circuits too.

An old image of what the updated ui looks like (the datatype text is no longer formatted like this anymore, it is now formatted as [datatype] list)
image

## Why It's Good For The Game

Composite lists are better because they let you more easily cast variables to specific types. The dispenser shell is also in this PR because I needed composite lists to properly do the dispenser shell.

Also fixes some general list-related bugs when it comes 

## Changelog

:cl:
expansion: Adds the dispenser shell, which accepts normal-sized items and lets you dispense them at any time.
expansion: Adds the For Each component and the Filter List component.
refactor: Refactored the circuit list datatype to support composite types (i.e. entity list, any list, number list), which are lists that can only contain values of the specified datatype. This also refactors anything that uses the list datatype to use composite lists.
qol: NTNet components now use composite lists to send and receive data
fix: Prevented people from infinitely making lists with the list literal component.
/:cl:

### CODE REVIEW

File: tgui/packages/tgui/interfaces/IntegratedCircuit/VariableMenu.js [hunk 1]
```diff
@@ -99,7 +99,7 @@ export class VariableMenu extends Component {
                             Drag me onto the circuit's grid
                             to make a setter for this variable`
                             }
-                            icon="hammer"
+                            icon="pen"
```
Comment (Colovorat) [inline tgui/packages/tgui/interfaces/IntegratedCircuit/VariableMenu.js:102]: maybe try download and upload signs or something?

File: code/modules/wiremod/components/list/list_literal.dm [hunk 2]
```diff
@@ -5,7 +5,7 @@
  */
 /obj/item/circuit_component/list_literal
 	display_name = "List Literal"
-	desc = "A component that returns the value of a list at a given index. Attack in hand to increase list size, right click to decrease list size."
+	desc = "A component that returns a list from whatever input you give it."
```
Comment (Colovorat) [inline code/modules/wiremod/components/list/list_literal.dm:8]: i know it's proper wording for it, but for non-coders Returns can mean bad thing or something, maybe "creates a list"?

File: tgui/packages/tgui/interfaces/IntegratedCircuit/VariableMenu.js (hunk 1 above)
Comment (Watermelon914) [reply tgui/packages/tgui/interfaces/IntegratedCircuit/VariableMenu.js:102]: pen is reasonable enough imo

File: code/modules/wiremod/components/list/list_literal.dm (hunk 2 above)
Comment (Watermelon914) [reply code/modules/wiremod/components/list/list_literal.dm:8]:
```suggestion
	desc = "A component that creates a list from whatever input you give it."
```

**Comment by FA120** [comment]
There's an oversight that harms using the crew computer component a lot...Since the lists that exist in the crew database table have indexes that are not numerical (name,life_status etc etc...) changing the index port in index list to number instead of any makes the values inaccessible

**Comment by Watermelon914** [comment]
> There's an oversight that harms using the crew computer component a lot...Since the lists that exist in the crew database table have indexes that are not numerical (name,life_status etc etc...) changing the index port in index list to number instead of any makes the values inaccessible

You can still use the `any` type.
Tables will be slightly more complicated to implement composite types into. They were made before option ports were available as a port so I'll look at refactoring them soon.

**Comment by FA120** [comment]
> > There's an oversight that harms using the crew computer component a lot...Since the lists that exist in the crew database table have indexes that are not numerical (name,life_status etc etc...) changing the index port in index list to number instead of any makes the values inaccessible
> 
> You can still use the `any` type. Tables will be slightly more complicated to implement composite types into. They were made before option ports were available as a port so I'll look at refactoring them soon so that they work better.

Actually,that doesn't fix it...My problem is not with the content of the list,my problem's that the index itself in the crew table isn't numeric..For exemple if you take one row from the crew table and try to apply list index on it,if you use numerics,you'd get 'name','life_status' 'job' strings instead of the actual value of name,of life_status etc...it wasn't a problem before since the index you had to use is that string itself,for exemple if you want the name of that row,you put 'name' in the index port instead of 1...Hope you understand me

**Comment by Watermelon914** [comment]
> > > There's an oversight that harms using the crew computer component a lot...Since the lists that exist in the crew database table have indexes that are not numerical (name,life_status etc etc...) changing the index port in index list to number instead of any makes the values inaccessible
> > 
> > 
> > You can still use the `any` type. Tables will be slightly more complicated to implement composite types into. They were made before option ports were available as a port so I'll look at refactoring them soon so that they work better.
> 
> Actually,that doesn't fix it...My problem is not with the content of the list,my problem's that the index itself in the crew table isn't numeric..For exemple if you take one row from the crew table and try to apply list index on it,if you use numerics,you'd get 'name','life_status' 'job' strings instead of the actual value of name,of life_status etc...it wasn't a problem before since the index you had to use is that string itself,for exemple if you want the name of that row,you put 'name' in the index port instead of 1...Hope you understand me

There's a special component that returns the list called Index Table

**Comment by FA120** [comment]
> > > > There's an oversight that harms using the crew computer component a lot...Since the lists that exist in the crew database table have indexes that are not numerical (name,life_status etc etc...) changing the index port in index list to number instead of any makes the values inaccessible
> > > 
> > > 
> > > You can still use the `any` type. Tables will be slightly more complicated to implement composite types into. They were made before option ports were available as a port so I'll look at refactoring them soon so that they work better.
> > 
> > 
> > Actually,that doesn't fix it...My problem is not with the content of the list,my problem's that the index itself in the crew table isn't numeric..For exemple if you take one row from the crew table and try to apply list index on it,if you use numerics,you'd get 'name','life_status' 'job' strings instead of the actual value of name,of life_status etc...it wasn't a problem before since the index you had to use is that string itself,for exemple if you want the name of that row,you put 'name' in the index port instead of 1...Hope you understand me
> 
> There's a special component that returns the list called Index Table

Again, the problem is not getting the list in itself... try to use index table to get one of the lists and then try to extract the name job life status or health from that list... You can't since the crew lists have very special index that instead of having a numerical index, its a string index, kinda like the map type in c++... lets suppose L is one of the lists that exist in the crew table, and lets assume it corresponds to someone named Hits-The-floor, L[1]='name' and L['name'] ='Hits-The-floor'

**Comment by Watermelon914** [comment]
> > snip
> Again, the problem is not getting the list in itself... try to use index table to get one of the lists and then try to extract the name job life status or health from that list... You can't since the crew lists have very special index that instead of having a numerical index, its a string index, kinda like the map type in c++... lets suppose L is one of the lists that exist in the crew table, and lets assume it corresponds to someone named Hits-The-floor, L[1]='name' and L['name'] ='Hits-The-floor'

You're talking about a row, which is different. I'll implement proper associative list indexing

**Comment by FA120** [comment]
> > > snip
> > 
> > 
> > Again, the problem is not getting the list in itself... try to use index table to get one of the lists and then try to extract the name job life status or health from that list... You can't since the crew lists have very special index that instead of having a numerical index, its a string index, kinda like the map type in c++... lets suppose L is one of the lists that exist in the crew table, and lets assume it corresponds to someone named Hits-The-floor, L[1]='name' and L['name'] ='Hits-The-floor'
> 
> You're talking about a row, which is different. I'll implement proper associative list indexing

That's what I meant but english isn't my First language...thanks

**Review by Mothblocks — CHANGES_REQUESTED** [review]
Mostly skimmed it, here's my reviews.

**Comment by Mothblocks** [comment]
lol

**Pull request merged**